### ISSUE TITLE
False negatives for D402

### ISSUE THREAD

**Issue by ns-cweber** [PyCQA/pydocstyle#284]
The following function fails the D402 check:

```py
def subclasses(cls):
    """Retrieve all subclasses (direct and transitive) from `cls`.

    :param cls: The class to search.
    :type cls: ``type``

    :returns: A generator of subclasses.
    :rtype: ``iter`` of ``type``
    """
    for subclass in cls.__subclasses__():
        for subsubclass in subclasses(subclass):
            yield subsubclass
        yield subclass
```

Presumably the D402 rule just sees the method name followed by an lparen?

**Comment by FarmerSez** [comment]
Indeed it does. This would also happen on later versions.

**Comment by Nurdok** [comment]
I think that validating that the content inside the parenthesis equals the parameter names might be a good fix.

**Comment by FarmerSez** [comment]
That might be problematic @Nurdok. What if you copied the signature to the docstring but had a typo? or changed the number of parameters and forgot to change the docstring? There will be a lot of false-negatives 😕.

**Comment by ns-cweber** [comment]
@FarmerSez Seems like you could get pretty close by matching on a valid Python function signature (function name, lparens, comma-delineated-args, rparens) and comparing the function names. This would weed out a lot of false negatives and also prevent the false positive you described without throwing a false negative for a function name followed by some things in parens.

**Comment by FarmerSez** [comment]
@ns-cweber that's true except if the parentheses contain only one word: `finds all subclasses (directly) using ....`, but I'm okay with that.

**Comment by chaoflow** [comment]
Another one to take into consideration is:

```python
def stat(path):
    """Wrap os.stat() for ease of mocking."""  # noqa: D402
    return os.stat(path)
```

I think we should match for valid function names only, not prefixed by a dot, directly followed by a opening parenthesis. This should also handle the original request on this issue.

**Comment by barakbeilin** [comment]
I'm working on a fix, in #450, having issues with adding a test in test.py for e.g.:
`@expect('D402: First line should not be the function\'s "signature"')
def foobar(foo, bar):
    """Signature: foobar(for,bar)."""`

The problem is to check a function with multiple arguments.

**Comment by seebi** [comment]
Just another false negative example we should avoid:

```
@click.group(cls=CmemcGroup)
def scheduler():
    """List, inspect, enable/disable or open scheduler(s).

    Schedulers execute workflows in specified intervals. They are identified
    with a SCHEDULERID. To get a list of existing schedulers, execute the
    list command or use tab-completion.
    """
```

removing "(s)" solved the issue.
